unistd/pwd.c

```diff
@@ -24,69 +24,94 @@
 
 struct passwd pwnam;
 char pw_name[NAME_MAX];
+char intstr[16];
 char pw_gecos[128];
 char pw_dir[PATH_MAX];
 char pw_shell[PATH_MAX];
 char pw_passwd[128];
 
 
-struct passwd *getpwnam(const char *name)
+static void readpwentry(FILE *fp, char *pwentry, unsigned int maxlgth)
 {
-	FILE *fp;
-	char buf[1024];
+	char c = 0;
+	unsigned int i = 0;
+	c = fgetc(fp);
+	while (c != ':' && c != '\n' && c != EOF) {
+		if (i < maxlgth - 1)
+			pwentry[i++] = c;
+		c = fgetc(fp);
+	}
+	pwentry[i] = 0;
+}
 
-	if ((fp = fopen(PASSWD_PATH, "r")) == NULL)
-		return NULL;
 
+static struct passwd *getpwby(FILE *fp, const char *name, uid_t *uid)
+{
 	pwnam.pw_name = pw_name;
-	pwnam.pw_gecos = pw_name;
 	pwnam.pw_dir = pw_dir;
 	pwnam.pw_gecos = pw_gecos;
 	pwnam.pw_shell = pw_shell;
 	pwnam.pw_passwd = pw_passwd;
 
-	while (fgets(buf, sizeof(buf), fp)) {
-		sscanf (buf, "%[^:]:%[^:]:%d:%d:%[^:]:%[^:]:%s\n",
-			pwnam.pw_name, pwnam.pw_passwd,
-			&pwnam.pw_uid, &pwnam.pw_gid,
-			pwnam.pw_gecos, pwnam.pw_dir, pwnam.pw_shell);
-		if (strcmp(pwnam.pw_name, name) == 0) {
-			fclose(fp);
-			return &pwnam;
+	do {
+		readpwentry(fp, pwnam.pw_name, NAME_MAX);
+		readpwentry(fp, pwnam.pw_passwd, sizeof(pw_passwd));
+		readpwentry(fp, intstr, sizeof(intstr));
+		pwnam.pw_uid = strtoul(intstr, NULL, 10);
+		readpwentry(fp, intstr, sizeof(intstr));
+		pwnam.pw_gid = strtoul(intstr, NULL, 10);
+		readpwentry(fp, pw_gecos, sizeof(pw_gecos));
+		readpwentry(fp, pwnam.pw_dir, PATH_MAX);
+		readpwentry(fp, pwnam.pw_shell, PATH_MAX);
+
+		/* two "if-s" for misra compliance */
+		if (name != NULL) {
+			if (strcmp(pwnam.pw_name, name) == 0)
+				break;
 		}
-	}
+		else {
+			if (pwnam.pw_uid == *uid)
+				break;
+		}
+
+	} while (!feof(fp));
+
+	if (feof(fp))
+		return NULL;
+	else
+		return &pwnam;
+}
+
+
+struct passwd *getpwnam(const char *name)
+{
+	FILE *fp;
+	struct passwd *ptr;
+
+	if (name == NULL)
+		return NULL;
+	if ((fp = fopen(PASSWD_PATH, "r")) == NULL)
+		return NULL;
+
+	ptr = getpwby(fp, name, NULL);
 	fclose(fp);
-	return NULL;
+
+	return ptr;
 }
 
 
 struct passwd *getpwuid(uid_t uid)
 {
 	FILE *fp;
-	char buf[1024];
+	struct passwd *ptr;
 
 	if ((fp = fopen(PASSWD_PATH, "r")) == NULL)
 		return NULL;
 
-	pwnam.pw_name = pw_name;
-	pwnam.pw_gecos = pw_name;
-	pwnam.pw_dir = pw_dir;
-	pwnam.pw_gecos = pw_gecos;
-	pwnam.pw_shell = pw_shell;
-	pwnam.pw_passwd = pw_passwd;
-
-	while (fgets(buf, sizeof(buf), fp)) {
-		sscanf (buf, "%[^:]:%[^:]:%d:%d:%[^:]:%[^:]:%s\n",
-			pwnam.pw_name, pwnam.pw_passwd,
-			&pwnam.pw_uid, &pwnam.pw_gid,
-			pwnam.pw_gecos, pwnam.pw_dir, pwnam.pw_shell);
-		if (pwnam.pw_uid == uid) {
-			fclose(fp);
-			return &pwnam;
-		}
-	}
+	ptr = getpwby(fp, NULL, &uid);
 	fclose(fp);
-	return NULL;
+
+	return ptr;
```

Comment:
How about doing the file opening in `getpwby`? Both function using it are opening the same file, so this can be included in `getpwby`.

Reply:
Moved opening to getpwby. It greatly reduced _real_ `getpwnam()`/`getpwby()` functions. Thanks!